File: Algorithms/Greedy/CoveringSegments.py

class Segment(object):

    def __init__(self, s, e):

        self.start = s
        self.end = e

def get_optimal_points(segments):

    segments.sort(key=lambda segment: segment.end)

    current = segments[0].end

    points = [current]

    for i in range(1, len(segments)):

        if current < segments[i].start or current > segments[i].end:

            current = segments[i].end

            points.append(current)

    return points

def main():

    segments = [
        Segment(4, 7),
        Segment(1, 3),
        Segment(2, 5),
        Segment(5, 6)
    ]

    for point in get_optimal_points(segments):

        print(point)

if __name__ == '__main__':
    main()

File: Algorithms/Greedy/CoveringSegments.js
"use strict";

class Segment {

    constructor(s, e) {

        this.start = s;
        this.end = e;

    }

}

let getOptimalPoints = (segments) => {

    segments.sort((a, b) => a.end - b.end);

    let current = segments[0].end;

    let points = [current];

    for (let i = 0; i < segments.length; i++) {

        if (current < segments[i].start || current > segments[i].end) {

            current = segments[i].end;

            points.push(current);

        }

    }

    return points;

}

let segments = [
    new Segment(4, 7),
    new Segment(1, 3),
    new Segment(2, 5),
    new Segment(5, 6)
];

for (let point of getOptimalPoints(segments)) {

    console.log(point);

}
